Various fixes for OpenSHMEM 1.5 support

dalcinl opened the pull request:
This PR is a bit large, but if you go commit by commit it should be relatively easy to review. Unfortunately, there aren't so many tests for the new 1.5 stuff, that probably explains  why so many issues got in unnoticed.

After the changes in this PR, OSHMPI passes all of my testing (from in-development Python wrappers no publicly available just yet but access available upon request).

Fixes #138
Fixes #139 
Fixes #140

dalcinl commented:
@hzhou Could you help me out to get this one in? I tried my best to write clean code. I'm not sure how to run the whitespace checker locally or trigger Jenkins to run it again after force-pushing to this PR.

dalcinl commented:
@yfguo You should most likely review all the changes I'm proposing in this PR.

dalcinl commented:
@hzhou Could you please re-run the GitHub Actions checks? The macOS runs failed for reasons unrelated to my changes (`brew install` failure, maybe now it will work).

hzhou reviewed (commented):
I gave a quick review. Looks good to me. I think it is okay to merge. @dalcinl If you rebase the PR I believe it will restart the action CI.

dalcinl commented on include/shmem.h.in.tpl:
```diff
@@ -102,10 +102,10 @@ typedef struct {
 
 typedef void* shmem_ctx_t;
 #define SHMEM_CTX_DEFAULT (shmem_ctx_t) 0x80000
+#define SHMEM_CTX_INVALID NULL
```
I'm going to add a new commit with this fix. I chose the value of `NULL` for symmetry with `SHMEM_TEAM_INVALID` (and not because of my advocacy for using NULL for invalid/null handles).

The reason I moved this down is because `SHMEM_CTX_INVALID` simply belongs together with `SHMEM_CTX_DEFAULT`, both are values for `shmem_ctx_t` handles, cf the declaration of the `SHMEM_TEAM_XXX` handles.

dalcinl commented:
@hzhou I've fixed the message of the first commit. Also, I've added a new one (the last one) that you did not had before, you may like to quickly review it (look at the message for rationale).

dalcinl commented:
> If you rebase the PR I believe it will restart the action CI.

That is not enough. For security reasons, workflows triggered by external contributors require approval to run.

dalcinl commented:
@hzhou @jeffhammond Folks, I would really like to get this one merged.

hzhou commented:
> @hzhou @jeffhammond Folks, I would really like to get this one merged.

Merged. Sorry for the neglect.
